R/method-introspect.R

```diff
@@ -31,30 +32,86 @@
 #' method(bizarro, object = 1)
 #' method(bizarro, new_S3_class("factor"))
 method <- function(generic, class = NULL, object = NULL) {
+  dispatch <- as_dispatch(generic, class = class, object = object)
+  .Call(method_, generic, dispatch, NULL)
+}
+
+#' Explain method dispatch
+#'
+#' @description
+#' `method_explain()` shows all possible methods that a call to a generic
+#' might use, which ones exist, and which one will actually be called.
+#'
+#' Note that method dispatch uses a string representation of each class in
+#' the class hierarchy. Each class system uses a slightly different convention
+#' to avoid ambiguity.
+#'
+#' * R7: `pkg::class` or `class`
+#' * S4: `S4/pkg::class` or `S4/class`
+#' * S3: `class`
+#'
+#' @inheritParams method
+#' @export
+#' @examples
+#' foo1 <- new_class("foo1")
+#' foo2 <- new_class("foo2", foo1)
+#'
+#' add <- new_generic("add", c("x", "y"))
+#' method(add, list(foo2, foo1)) <- function(x, y) c(2, 1)
+#' method(add, list(foo1, foo1)) <- function(x, y) c(1, 1)
+#'
+#' method_explain(add, list(foo2, foo2))
+method_explain <- function(generic, class = NULL, object = NULL) {
+  dispatch <- as_dispatch(generic, class = class, object = object)
+
+  grid <- as.matrix(rev(do.call("expand.grid", rev(dispatch))))
+  colnames(grid) <- generic@dispatch_args
+
+  names <- paste0("[", grid, "]")
```

Comment:
You like `[]` better than `<>`? It seems like we are printing types here, and `<>` felt more consistent